### PR TITLE
Factor out cert handler

### PR DESCRIPTION
**Pull request by AlCutter**
This one is mostly just moving stuff around; pulling out the CT/certificate specific stuff from `handler.cc` and putting it in a subclass, then there's also the changes to support registering these handler subclasses with the `Server` rather than the server creating the handler instance(s).

I'm not super happy with the current state of the Server interface, but I think that's a job for another day.

### CODE REVIEW

**Comment by Martin2112** [comment]
This looks related to my WIP in #878.

Martin

File: cpp/server/handler.cc [hunk 1]
```diff
@@ -1,237 +1,78 @@
...
-using cert_trans::CertChain;
-using cert_trans::CertChecker;
-using cert_trans::Counter;
-using cert_trans::HttpHandler;
-using cert_trans::JsonOutput;
-using cert_trans::Latency;
-using cert_trans::LoggedEntry;
-using cert_trans::Proxy;
-using cert_trans::ScopedLatency;
+#include "server/handler.h"
```
Comment (pphaneuf) [inline cpp/server/handler.cc]: What the heck happened to this file? Its header should be first, and generally, this reshuffling is making it rather difficult to see what's going on with *this* PR...

**Comment by pphaneuf** [comment]
Could you review #995, and rebase on top once I've merged it? :stuck_out_tongue:

File: cpp/server/handler.cc (hunk 1 above)
Comment (AlCutter) [inline cpp/server/handler.cc]: Ah, this is because this file briefly turned into a template (remember?) and then back out to a regular class.
I've moved the class header back up top, though.

**Comment by AlCutter** [comment]
Ping

File: cpp/server/handler.cc [hunk 2]
```diff
@@ -3,160 +3,77 @@
...
-using cert_trans::Latency;
-using cert_trans::LoggedEntry;
-using cert_trans::Proxy;
-using cert_trans::ScopedLatency;
+#include "util/libevent_wrapper.h"
+#include "util/sync_task.h"
+#include "util/task.h"
+
+
+DEFINE_int32(max_leaf_entries_per_response, 1000,
```
Comment (pphaneuf) [inline cpp/server/handler.cc]: Did those change at all, or did they just migrate for the coming winter months? :wink: 

I was kind of hoping for "diff minimisation". :smiley:

File: cpp/server/handler.cc [hunk 3]
```diff
@@ -3,160 +3,77 @@
...
+             "maximum number of entries to put in the response of a "
+             "get-entries request");
+DEFINE_int32(staleness_check_delay_secs, 5,
+             "number of seconds between node staleness checks");
+
+
+namespace cert_trans {
```
Comment (pphaneuf) [inline cpp/server/handler.cc]: I think if you moved lines 26-33 to line 47, it'd make more sense? Small diff *and* reasonable thing to do? (it might be reasonable now, but you're making my pretty little head hurt!)

Comment (AlCutter) [inline cpp/server/handler.cc]: moved

Comment (pphaneuf) [inline cpp/server/handler.cc]: All right, thanks! I can now see that many relevant things *did* happen within that noise! :smiley:

File: cpp/server/handler.cc [hunk 4]
```diff
@@ -3,160 +3,77 @@
 #include <algorithm>
 #include <event2/buffer.h>
```
Comment (pphaneuf) [inline cpp/server/handler.cc]: I *think* this one might not be needed either, with `ExtractChain` having moved? Not 100% sure about `event2/http.h`, though.

Comment (AlCutter) [inline cpp/server/handler.cc]: Dinged those two (on a related note, we should run IWYU over the codebase rather than trying to do this manually.)

**Comment by AlCutter** [comment]
Ping

File: cpp/log/cert_submission_handler.cc [hunk 5]
```diff
@@ -76,7 +76,8 @@ bool CertSubmissionHandler::X509ChainToEntry(const CertChain& chain,
 }
 
 Status CertSubmissionHandler::ProcessX509Submission(CertChain* chain,
-                                                    LogEntry* entry) {
+                                                    LogEntry* entry) const {
+  entry->set_type(ct::X509_ENTRY);
```
Comment (pphaneuf) [inline cpp/log/cert_submission_handler.cc]: This here, was this a separate bug?!?

File: cpp/log/frontend_test.cc [hunk 6]
```diff
@@ -162,8 +163,11 @@ TYPED_TEST(FrontendTest, TestSubmitValid) {
   CertChain chain(this->leaf_pem_);
   EXPECT_TRUE(chain.IsLoaded());
 
+  LogEntry entry;
   SignedCertificateTimestamp sct;
-  EXPECT_OK(this->frontend_.QueueX509Entry(&chain, &sct));
+  EXPECT_OK(this->frontend_.QueueProcessedEntry(
+      this->submission_handler_.ProcessX509Submission(&chain, &entry), entry,
```
Comment (pphaneuf) [inline cpp/log/frontend_test.cc]: Passing the same `entry` twice, with one of them looking "by value" (yeah, I know, thank you C++!), that looked rather alarming.

To my eye, it really looks like `ProcessX509Submission` and `ProcessPreCertSubmission` want to return a `StatusOr<LogEntry>`...

Comment (AlCutter) [inline cpp/log/frontend_test.cc]: Yep, that's a good idea and actually fits very well with what `QueueProcessedEntry` really does, I just kept things the same 'shape' as they were previously, but we can do this change in a separate PR.

File: cpp/server/handler.h [hunk 7]
```diff
@@ -32,21 +33,27 @@ class ThreadPool;
...
 
   void Add(libevent::HttpServer* server);
 
- private:
+  void SetOutput(JsonOutput* json_output);
+  void SetProxy(Proxy* proxy);
+
+ protected:
+  // Implemented by subclasses which want to add their own extra http handlers.
+  virtual void AddHandlers(libevent::HttpServer* server) = 0;
```
Comment (pphaneuf) [inline cpp/server/handler.h]: I'm still fuzzy on why it's not just `HttpHandler::Add` that shouldn't just become virtual?

File: cpp/server/certificate_handler.cc [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+
+CertificateHttpHandler::CertificateHttpHandler(
+    LogLookup<LoggedEntry>* log_lookup,
+    const ReadOnlyDatabase<LoggedEntry>* db,
+    const ClusterStateController<LoggedEntry>* controller,
+    const CertChecker* cert_checker, Frontend* frontend, ThreadPool* pool,
+    libevent::Base* event_base)
+    : HttpHandler(log_lookup, db, controller, pool, event_base),
+      cert_checker_(cert_checker),
```
Comment (pphaneuf) [inline cpp/server/certificate_handler.cc]: Some `CHECK_NOTNULL`?

**Comment by pphaneuf** [comment]
Review done! **I can see!**

File: cpp/server/handler.h (hunk 7 above)
Comment (AlCutter) [inline cpp/server/handler.h]: I could do, it's just that I figured that all subclasses will want to map the `get-entries`, `get-proof-by-hash`, `get-sth`, and `get-sth-consistency` handlers, so may as well just bake them in here.
If `Add` is virtual then the subclass has to either add them all itself, or include a call to the superclass `Add` implementation.  The former is a bunch of extra boilerplate for no real reason, and the latter would be ok but costs an extra line and could also lead to confusing "why are only *some* handlers missing" bugs.

File: cpp/server/certificate_handler.cc (hunk 8 above)
Comment (AlCutter) [inline cpp/server/certificate_handler.cc]: `HttpHandler` checks all of its params for not-nullness, `frontend` is permitted to be null, but I've added a check for `cert_checker`

**Comment by pphaneuf** [comment]
LGTM
